rosdiagnostic/src/rosdiagnostic/rosdiagnostic.py

```diff
@@ -0,0 +1,234 @@
+    """
+    Make sure that master is available
+    :raises: :exc:`ROSTopicException` If unable to successfully communicate with master
+    """
+    try:
+        rosgraph.Master('/rostopic').getPid()
+    except socket.error:
+        raise ROSTopicIOException("Unable to communicate with master!")
+
+
+class RosDiagnostics(object):
+
+    DIAG_MSG_LVL_STR_NOCOLOR = {
+        DiagnosticStatus.OK:    ' OK  ',
+        DiagnosticStatus.WARN:  'WARN ',
+        DiagnosticStatus.ERROR: 'ERROR',
+        DiagnosticStatus.STALE: 'STALE',
+    }
+
+    DIAG_MSG_LVL_STR_COLOR = {
+        DiagnosticStatus.OK:    '\033[92m OK  \033[0m',
+        DiagnosticStatus.WARN:  '\033[93mWARN \033[0m',
+        DiagnosticStatus.ERROR: '\033[91mERROR\033[0m',
+        DiagnosticStatus.STALE: '\033[94mSTALE\033[0m',
+    }
+
+    def __init__(self, options):
+        self._lvl_strings_map = self.DIAG_MSG_LVL_STR_NOCOLOR if options.nocolor else self.DIAG_MSG_LVL_STR_COLOR
+        self._options = options
+        rospy.Subscriber('{}/diagnostics_agg'.format(options.ns), DiagnosticArray, self._diag_callback, queue_size=1)
+
+    def _get_ns(self, name):
+        # Split namespace by slashes, return the base namespace.
+        return '/'.join(name.split('/')[:-1])
+
+    def _get_non_leaf_statuses(self, statuses):
+        return [self._get_ns(s.name) for s in statuses]
+
+    def _get_leaf_statuses(self, statuses):
+        parent_namespaces = self._get_non_leaf_statuses(statuses)
+        return [s for s in statuses if s.name not in parent_namespaces]
+
+    def _print_begin_banner(self, msg):
+        dt = datetime.fromtimestamp(msg.header.stamp.to_sec(), tzlocal())
+        print('=====================================================================')
+        print('Diagnostic generated on: {}'.format(dt))
+        print('---------------------------------------------------------------------')
+
+    def _print_end_banner(self, msg):
+        print('=====================================================================')
+        if not self._options.follow:
+            rospy.signal_shutdown('Run complete')
+
+    def _diag_callback(self, msg):
+        prog = re.compile(self._options.filter)
+        self._print_begin_banner(msg)
+        for diag in sorted(self._get_leaf_statuses(msg.status), key=lambda d: d.level, reverse=True):
+            if prog.search(diag.name):
+                self._print_status(diag)
+        self._print_end_banner(msg)
+
+    def _print_status(self, diag):
+        if self._options.min_level <= diag.level:
+            print('[ {} ] {} - {}'.format(self._lvl_strings_map[diag.level], diag.name, diag.message))
+            if self._options.detail:
+                print('    hardware_id: {}'.format(diag.hardware_id))
+                for kv in diag.values:
+                    print('    - {}: {}'.format(kv.key, kv.value))
+
+
+def _rosdiagnostic_cmd_echo(argv):
+    parser = argparse.ArgumentParser(description='CPR ROS Diagnostics')
```

Comment:
I think the description here should be `ROS Diagnostics` or `ROS Diagnostic Viewer` now that this is going to be an upstream tool.